add VerifiedMaintainer field to Addons and display a warning if empty

inifares23lab opened the pull request:
The Addon type now has another string Field: VerifiedMaintainer
The new field is added to the NewAddon function and to the Addons map.

When enabling an addon if the VerifiedMaintainer field is empty a warning message is displayed:

! The <ADDON_NAME> addon doesn't have a verified maintainer.
Since for now I just added an empty string for every addon, it will always display the warning when enabling

fixes #14493

inifares23lab commented:
Hi @spowelljr
apologies
formatted and pushed now

spowelljr commented:
Hi @inifares23lab, thanks for formatting the code.

Could you wrap all `Maintainers` that aren't `Google` or `Kubernetes` with `3rd party(X)`.

For example `Portainer.io` to `3rd party (Portainer.io)`

Could you also add `3rd party (Gluster)` as the maintainer for `storage-provisioner-gluster`.

If the `Maintainer` is `Kubernetes` or `Google` could you make the message be the following:
```
<addon_name> is an addon maintained by <maintainer> for any concerns contact minikube on GitHub.
You can view the list of minikube maintainers at: https://github.com/kubernetes/minikube/blob/master/OWNERS
```

For the rest of addons could you update your message to:
```
<addon_name> is A 3rd party addon and not maintained or verified by minikube maintainers, enable at your own risk.
```

And for 3rd party addons, if there's a verified maintainer for the addon could you also add the following line:
```
<addonn_name> is maintained by <maintainer> for any concerns contact <verified_maintaer> on GitHub.
```

inifares23lab commented:
Hello

- updated messages and handled the 3 different cases
- updated 3rd party maintainers and fixed some typos

examples:
💡  auto-pause is an addon maintained by Google. For any concerns contact minikube on GitHub.
    You can view the list of minikube maintainers at: https://github.com/kubernetes/minikube/blob/master/OWNERS

❗  ambassador is A 3rd party addon and not maintained or verified by minikube maintainers, enable at your own risk.

💡  metallb is maintained by 3rd party (MetalLB) for any concerns contact MOCK-VERIFI
ED-MAINTAINER on GitHub.

spowelljr commented on cmd/minikube/cmd/config/enable.go:
```diff
@@ -52,6 +53,21 @@ var addonsEnableCmd = &cobra.Command{
+			maintainer := addonBundle.Maintainer
+			if maintainer == "Google" || maintainer == "Kubernetes" {
+				out.Styled(style.Tip, `{{.addon}} is an addon maintained by {{.maintainer}}. For any concerns contact minikube on GitHub.
+    You can view the list of minikube maintainers at: https://github.com/kubernetes/minikube/blob/master/OWNERS`,
```
```suggestion
You can view the list of minikube maintainers at: https://github.com/kubernetes/minikube/blob/master/OWNERS`,
```

spowelljr commented on cmd/minikube/cmd/config/enable.go:
```diff
@@ -52,6 +53,21 @@ var addonsEnableCmd = &cobra.Command{
+				out.Styled(style.Tip, `{{.addon}} is maintained by {{.maintainer}} for any concerns contact {{.verified_maintainer}} on GitHub.`,
+					out.V{"addon": addon, "maintainer": maintainer, "verified_maintainer": verifiedMaintainer})
+			} else {
+				out.Styled(style.Warning, `{{.addon}} is A 3rd party addon and not maintained or verified by minikube maintainers, enable at your own risk.`,
```
```suggestion
				out.Styled(style.Warning, `{{.addon}} is a 3rd party addon and not maintained or verified by minikube maintainers, enable at your own risk.`,
```

spowelljr commented on cmd/minikube/cmd/config/enable.go:
```diff
@@ -52,6 +53,21 @@ var addonsEnableCmd = &cobra.Command{
+					out.V{"addon": addon, "maintainer": maintainer})
+			} else if verifiedMaintainer := addonBundle.VerifiedMaintainer; verifiedMaintainer != "" {
+				out.Styled(style.Tip, `{{.addon}} is maintained by {{.maintainer}} for any concerns contact {{.verified_maintainer}} on GitHub.`,
+					out.V{"addon": addon, "maintainer": maintainer, "verified_maintainer": verifiedMaintainer})
```
```suggestion
					out.V{"addon": addon, "maintainer": maintainer, "verifiedMaintainer": verifiedMaintainer})
					
```
Can we keep normal Go casing here

inifares23lab replied:
of course

spowelljr commented on cmd/minikube/cmd/config/enable.go:
```diff
@@ -52,6 +53,21 @@ var addonsEnableCmd = &cobra.Command{
+				out.Styled(style.Tip, `{{.addon}} is an addon maintained by {{.maintainer}}. For any concerns contact minikube on GitHub.
+You can view the list of minikube maintainers at: https://github.com/kubernetes/minikube/blob/master/OWNERS`,
+					out.V{"addon": addon, "maintainer": maintainer})
+			} else if verifiedMaintainer := addonBundle.VerifiedMaintainer; verifiedMaintainer != "" {
+				out.Styled(style.Tip, `{{.addon}} is maintained by {{.maintainer}} for any concerns contact {{.verifiedMaintainer}} on GitHub.`,
+					out.V{"addon": addon, "maintainer": maintainer, "verifiedMaintainer": verifiedMaintainer})
```
```suggestion
			} else if addonBundle.VerifiedMaintainer != "" {
				out.Styled(style.Tip, `{{.addon}} is maintained by {{.maintainer}} for any concerns contact {{.verifiedMaintainer}} on GitHub.`,
					out.V{"addon": addon, "maintainer": maintainer, "verifiedMaintainer": addonBundle.VerifiedMaintainer})
```

The assigning within the if statement is a bit long and makes it a bit unreadable, I feel doing a direct comparison without assigning a new variable would be better

spowelljr reviewed (changes requested):
The Google/Kubernetes and no verified maintainer case is perfect.

But if it's a 3rd party and has a verified maintainer it should have both messages:
```
<addon_name> is A 3rd party addon and not maintained or verified by minikube maintainers, enable at your own risk.
<addon_name> is maintained by <maintainer> for any concerns contact <verified_maintaer> on GitHub.
```

That ways they're notified it's a 3rd party addon and to enable at their own risk, but also pointed to someone if they have troubles.

inifares23lab commented:
Hi @spowelljr

I agree that displaying both messages is more clear for the user, however I have one concern:

- the first message also states that the addon is not verified by minikube maintainers even if in some cases it has a VerifiedMaintainer


It may be better to further customize the messages for the two cases:

1. 3rd party without verified maintainer(same as before):
> <addon_name> is A 3rd party addon and not maintained or verified by minikube maintainers, enable at your own risk.

2. 3rd party with a verified maintainer(add the first message but omit "or verified"):
> <addon_name> is A 3rd party addon and not maintained by minikube maintainers, enable at your own risk.
> <addon_name> is maintained by <maintainer> for any concerns contact <verified_maintaer> on GitHub.


Let me know if it makes sense to you
thanks

spowelljr commented:
Hi @inifares23lab, I know in the issue I originally mentioned a verified maintainer but that was just something I quickly jotted down and should have named it something else.

The field can be called anything, `GitHubMaintainer` or something like that, there's really no verifying the user themselves. Verified is related to the addon itself though, and any addon that's not maintained by Kubernetes or Google isn't verified.

So the message:
`<addon_name> is A 3rd party addon and not maintained or verified by minikube maintainers, enable at your own risk.`
Seems fine to me for both cases because there is no verified maintainer.

inifares23lab commented:
understood, thank you for clarifying.
I applied the requested changes

spowelljr commented on pkg/minikube/assets/addons.go:
```diff
@@ -440,7 +442,7 @@ var Addons = map[string]*Addon{
 			"logviewer-rbac.yaml",
 			"0640"),
-	}, false, "logviewer", "", "", map[string]string{
+	}, false, "logviewer", "3rd party (ivans3)", "", "", map[string]string{
```
```suggestion
	}, false, "logviewer", "3rd party (unknown)", "", "", map[string]string{
```

spowelljr commented on pkg/minikube/assets/addons.go:
```diff
@@ -145,21 +147,21 @@ var Addons = map[string]*Addon{
 			"pod-security-policy.yaml",
 			"0640"),
-	}, false, "pod-security-policy", "", "", nil, nil),
+	}, false, "pod-security-policy", "3rd party (pod-security-policy)", "", "", nil, nil),
```
```suggestion
	}, false, "pod-security-policy", "3rd party (unknown)", "", "", nil, nil),
```

spowelljr commented on pkg/minikube/assets/addons.go:
```diff
@@ -386,7 +388,7 @@ var Addons = map[string]*Addon{
 			"patch-coredns-job.yaml",
 			"0640"),
-	}, false, "registry-aliases", "", "", map[string]string{
+	}, false, "registry-aliases", "3rd party (registry-aliases)", "", "", map[string]string{
```
```suggestion
	}, false, "registry-aliases", "3rd party (unknown)", "", "", map[string]string{
```

spowelljr reviewed (changes requested):
Looks good! Just a few last comments

spowelljr commented:
/ok-to-test

spowelljr reviewed (approved):
Looks good, thanks for your work on this!
